util/stream.go

```diff
@@ -156,40 +136,76 @@ func (m *MessageStream) inbound() {
+		for buf.Len() >= 4 {
+			if totalLen == 0 {
+				// Read the OpenFlow message length.
+				msgType := int(buf.Bytes()[1])
+				totalLen = int(binary.BigEndian.Uint16(buf.Bytes()[2:4]))
+				// msgType == openflow15.Type_Experimenter
+				if msgType == 4 {
+					// The minimum length of a valid VendorHeader message is 16 bytes.
+					if buf.Len() < 16 {
+						break
+					}
+					experimenterType := binary.BigEndian.Uint32(buf.Bytes()[12:])
+					// experimenterType == openflow15.Type_PacketIn2
+					if experimenterType == 30 {
+						// The first 4 byte of a PacketIn2 message is needed to check the packet length.
+						if buf.Len() < 20 {
+							break
+						}
+						// According to OVS implementation, the first property of a PacketIn2 message is NXPINT_PACKET.
+						pktProp := int(binary.BigEndian.Uint16(buf.Bytes()[16:]))
+						// pkgProp == openflow15.NXPINT_PACKET
+						if pktProp == 0 {
+							pktLength := int(binary.BigEndian.Uint16(buf.Bytes()[18:]))
+							if totalLen < pktLength {
+								totalLen += 1 << 16
+								klog.V(2).InfoS("Oversize packet detected: OpenFlow PacketIn message length overflowed", "message_length", totalLen)
+								// Reset the VendorHeader.Vendor field to mark the message is oversize.
+								binary.BigEndian.PutUint32(buf.Bytes()[8:12], 0x10002320)
+							}
+						}
+					}
 				}
-				continue
-			}
-			if msgLen > 0 {
-				buf.WriteByte(tmpBuf[i])
-				msgLen = msgLen - 1
-				if msgLen == 0 {
-					hdr = 0
-					m.dispatchMessage(buf)
-					buf = <-m.pool.Empty
+				klog.V(5).InfoS("Expected OpenFlow message", "length", totalLen)
+
+				// Return error if the message is shorter than the minimum length of a standard OpenFlow message.
+				if totalLen < 8 {
+					klog.Error("Buffer too small to parse OpenFlow messages")
+					err = fmt.Errorf("invalid message with length %d is received", totalLen)
+					m.Error <- err
+					m.Shutdown <- true
+					return
 				}
-				continue
 			}
-		}
-	}
-}
 
-// Dispatch the message to streamWorker according to Xid in the message Header
-func (m *MessageStream) dispatchMessage(b *bytes.Buffer) {
-	msgBytes := b.Bytes()
-	if len(msgBytes) < 8 {
-		klog.Error("Buffer too small to parse OpenFlow messages")
-		return
+			// If the openflow message is not completed, continue reading from the connection.
+			if buf.Len() < totalLen {
+				break
+			}
+
+			// Dispatch the message bytes to worker.
+			msgBytes := make([]byte, totalLen)
+			if _, err = buf.Read(msgBytes); err != nil {
+				// io.EOF is the only error returned by buf.Read.
+				klog.ErrorS(err, "Failed to read bytes from buffer")
+				m.Error <- err
+				m.Shutdown <- true
+				return
+			}
+
+			klog.V(5).InfoS("Received message", "message_length", totalLen, "buffer_length", len(msgBytes))
+			xid := binary.BigEndian.Uint32(msgBytes[4:])
+			workerKey := int(xid % uint32(len(m.workers)))
+			m.workers[workerKey].Full <- bytes.NewBuffer(msgBytes)
```

Comment:
I think using `Full` as the field name now that you got rid of the buffer pool (and its `Empty` channel counterpart) is misleading

Reply:
Maybe just `InputCh`. I notice it is not buffered (I don't know if this was intentional), so it will block if the worker is already processing a message, and the name `Queue` is not really appropriate.